### PR TITLE
CDAP-3605 Intercept entitiy deletion to also remove metadata.

### PR DESCRIPTION
**Pull request by henry-cask**
Intercept the entity deletion of App, Program, Dataset, and Stream so it also remove the corresponding business metadata (property and tags).

### CODE REVIEW

**Comment by henry-cask** [comment]
Build:
http://builds.cask.co/browse/CDAP-DUT2851-2

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java [hunk 1]
```diff
@@ -688,6 +693,8 @@ void deleteApp(Id.Application appId, ApplicationSpecification spec) throws Excep
 
     store.removeApplication(appId);
 
+    deleteAppBusinessMetadata(appId);
```
Comment (sreevatsanraman) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java]: this should be before we remove app from MDS.

**Comment by sreevatsanraman** [comment]
No test case for this PR. Don't merge without adding one.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java [hunk 2]
```diff
@@ -695,6 +702,65 @@ void deleteApp(Id.Application appId, ApplicationSpecification spec) throws Excep
...
+    for (ProgramType programType : ProgramType.values()) {
+      deleteProgramBusinessMetadataForApp(programType, appId, appSpec);
+    }
+  }
+
+  /**
+   *  Delete the business metadata for a program type in an application.
+   *  TODO: Need to remove this we support prefix search of metadata type.
+   */
+  private void deleteProgramBusinessMetadataForApp(ProgramType programType, Id.Application appId,
```
Comment (sreevatsanraman) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java]: This code is brittle. Everytime we add a new program type, we need to remember to add the program type here to delete. Why not delete the Business metadata based on a prefix key for app.

**Comment by sreevatsanraman** [comment]
This PR doesn't take into account when we remove programs on re-deploying an app. Look at `DeletedProgramHandlerStage`

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java (hunk 2 above)
Comment (sreevatsanraman) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java]: nvm i see a comment about this. Let's do this in 3.2

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java [hunk 3]
```diff
@@ -695,6 +702,65 @@ void deleteApp(Id.Application appId, ApplicationSpecification spec) throws Excep
...
+        LOG.info("Unable to remove business metadata for unsupported program type: " + programType.name());
+        break;
+    }
+
+    if (programNames == null) {
+      LOG.debug("No program metadata removed for application {}", appId.toString());
+      return;
+    }
+
+    for (String programName : programNames) {
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java]: Doesn't this also require notifications of the change? How are the notifications sent? Perhaps all meta data notifications should go through the metadata service?

**Comment by henry-cask** [comment]
Break lots of unit tests. Need to revisit this.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java (hunk 3 above)
Comment (henry-cask) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java]: Are we doing similar things for programs related changes? This is not actually change the app metastore but separate dataset that store metadata for programs.

**Pull request closed without merging**